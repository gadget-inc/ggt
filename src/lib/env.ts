import type { Paths } from "env-paths";
import envPaths from "env-paths";

/**
 * Captures the name and nature of the environment
 */
export class Env {
  static get value(): string {
    return process.env["GGT_ENV"] || "production";
  }

  static get paths(): Paths {
    return envPaths("gadget", { suffix: this.productionLike ? "" : this.value });
  }

  static get developmentLike(): boolean {
    return this.value.startsWith("development");
  }

  static get testLike(): boolean {
    return this.value.startsWith("test");
  }

  static get developmentOrTestLike(): boolean {
    return this.developmentLike || this.testLike;
  }

  static get productionLike(): boolean {
    return this.value.startsWith("production");
  }
}
