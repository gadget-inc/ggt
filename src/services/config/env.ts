import process from "node:process";

/**
 * Captures the name and nature of the environment
 */
export const env = {
  get value(): string {
    return process.env["GGT_ENV"] || "production";
  },

  get productionLike(): boolean {
    return !this.developmentOrTestLike;
  },

  get developmentLike(): boolean {
    return this.value.startsWith("development");
  },

  get testLike(): boolean {
    return this.value.startsWith("test");
  },

  get developmentOrTestLike(): boolean {
    return this.developmentLike || this.testLike;
  },
};
