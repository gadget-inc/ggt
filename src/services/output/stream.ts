import process from "node:process";
import { env } from "../config/env.js";
import { isObject } from "../util/is.js";

/**
 * A wrapper around process.stdout and process.stderr that allows us to mock out the streams for testing.
 *
 * @see https://github.com/oclif/core/blob/16139fe8a7f991b4b446a1599ab63f15d9809b8e/src/cli-ux/stream.ts
 */
export class Stream {
  /**
   * Indicates whether the last line that was written was empty. This is
   * used to prevent writing more than one empty line in a row.
   */
  private _lastLineWasEmpty = true;

  constructor(public channel: "stdout" | "stderr") {
    process[this.channel].on("error", (err: unknown) => {
      if (isObject(err) && "code" in err && err.code === "EPIPE") {
        return;
      }
      throw err;
    });
  }

  get isTTY(): boolean {
    return process[this.channel].isTTY;
  }

  getWindowSize(): number[] {
    return process[this.channel].getWindowSize();
  }

  write(str: string): void {
    // remove duplicate empty lines
    while (str.startsWith("\n\n")) {
      str = str.slice(0, -1);
    }

    while (str.endsWith("\n\n")) {
      str = str.slice(0, -1);
    }

    str = str.replaceAll(/\n\n+/g, "\n\n");

    if (this._lastLineWasEmpty) {
      // we just printed an empty line, so don't print another one
      while (str.startsWith("\n")) {
        str = str.slice(1);
      }
    }

    if (str === "") {
      // nothing to print
      return;
    }

    // remember if the last line was empty
    this._lastLineWasEmpty = str === "\n";

    this._write(str);
  }

  on(event: string, listener: (...args: unknown[]) => void): this {
    process[this.channel].on(event, listener);
    return this;
  }

  once(event: string, listener: (...args: unknown[]) => void): this {
    process[this.channel].once(event, listener);
    return this;
  }

  private _write(data: string): boolean {
    if (env.testLike) {
      // use console.log/error in tests since vitest doesn't display
      // process.stdout/stderr correctly; also, remove trailing newline
      // since console.log/error adds one.
      data = data.replace(/\n$/, "");
      if (this.channel === "stdout") {
        console.log(data);
      } else {
        console.error(data);
      }
      return true;
    }

    return process[this.channel].write(data);
  }
}

export const stdout = new Stream("stdout");
export const stderr = new Stream("stderr");
