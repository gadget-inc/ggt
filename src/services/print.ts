import ansiColors from "ansi-colors";
import chalkTemplate from "chalk-template";
import CliTable3, { type TableConstructorOptions } from "cli-table3";
import levenshtein from "fast-levenshtein";
import assert from "node:assert";
import process from "node:process";
import stripAnsi from "strip-ansi";
import { dedent } from "ts-dedent";
import { config } from "./config.js";
import { isObject, isString } from "./is.js";
import { createLogger, type Logger } from "./log.js";

export const color = ansiColors;
export const symbol = ansiColors.symbols;

/**
 * A wrapper around process.stdout and process.stderr that allows us to mock out the streams for testing.
 *
 * @see https://github.com/oclif/core/blob/16139fe8a7f991b4b446a1599ab63f15d9809b8e/src/cli-ux/stream.ts
 */
export class Stream {
  private _log?: Logger;

  public constructor(public channel: "stdout" | "stderr") {
    process[this.channel].on("error", (err: unknown) => {
      if (isObject(err) && "code" in err && err.code === "EPIPE") {
        return;
      }
      throw err;
    });
  }

  public get isTTY(): boolean {
    return process[this.channel].isTTY;
  }

  public getWindowSize(): number[] {
    return process[this.channel].getWindowSize();
  }

  public write(data: string): boolean {
    if (config.debug) {
      this._log ??= createLogger(this.channel);
      for (const line of stripAnsi(data).split("\n")) {
        this._log.debug(line as Lowercase<string>);
      }
      return true;
    } else {
      return process[this.channel].write(data);
    }
  }

  public on(event: string, listener: (...args: unknown[]) => void): this {
    process[this.channel].on(event, listener);
    return this;
  }

  public once(event: string, listener: (...args: unknown[]) => void): this {
    process[this.channel].once(event, listener);
    return this;
  }
}

export const stdout = new Stream("stdout");
export const stderr = new Stream("stderr");

export const sprint = (template: TemplateStringsArray | string, ...values: unknown[]): string => {
  let content = template;
  if (!isString(content)) {
    content = chalkTemplate(content, ...values);
  }
  return dedent(content);
};

export const sprintln = (template: TemplateStringsArray | string, ...values: unknown[]): string => {
  return sprint(template, ...values) + "\n";
};

export const sprintlns = (template: TemplateStringsArray | string, ...values: unknown[]): string => {
  return "\n" + sprintln(template, ...values);
};

export const print = (template: TemplateStringsArray | string, ...values: unknown[]): void => {
  stdout.write(sprint(template, ...values));
};

export const println = (template: TemplateStringsArray | string, ...values: unknown[]): void => {
  stdout.write(sprintln(template, ...values));
};

export const printlns = (template: TemplateStringsArray | string, ...values: unknown[]): void => {
  stdout.write(sprintlns(template, ...values));
};

/**
 * EXAMPLE:
 *    "top-left": "╔",    top: "═",    "top-mid": "╤",    "top-right": "╗",
 *          left: "║",                    middle: "│",          right: "║",
 *    "left-mid": "╟",    mid: "─",    "mid-mid": "┼",    "right-mid": "╢",
 * "bottom-left": "╚", bottom: "═", "bottom-mid": "╧", "bottom-right": "╝",
 */
export const printTable = ({
  rows,
  ...options
}: TableConstructorOptions & {
  rows: string[][];
}): void => {
  const table = new CliTable3({
    ...options,
    style: { head: [], border: [], ...options.style },
    // prettier-ignore
    chars: {
      "top-left": "",    top: "",    "top-mid": "",    "top-right": "",
      "left-mid": "",    mid: "",    "mid-mid": "",    "right-mid": "",
            left: "",                   middle: "",          right: "",
   "bottom-left": "", bottom: "", "bottom-mid": "", "bottom-right": "",
      ...options.chars
  },
  });

  table.push(...rows);

  let output = table.toString();
  if (!options.head || options.head.length === 0) {
    // cli-table3 adds a single space to the first row when there are no
    // headers, so we remove it here
    output = output.slice(1);
  }

  println(output);
};

export const sortBySimilarity = (input: string, options: Iterable<string>): [closest: string, ...sorted: string[]] => {
  const strings = Array.from(options);
  assert(strings.length > 0, "options must not be empty");
  return strings.sort((a, b) => levenshtein.get(a, input) - levenshtein.get(b, input)) as [string, ...string[]];
};
