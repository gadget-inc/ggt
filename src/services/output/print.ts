import type { Options as BoxenOptions } from "boxen";
import boxen from "boxen";
import chalkTemplate from "chalk-template";
import CliTable3 from "cli-table3";
import { dedent } from "ts-dedent";
import { config } from "../config/config.js";
import { isString } from "../util/is.js";
import { defaults } from "../util/object.js";
import type { Field } from "./log/field.js";
import { formatPretty } from "./log/format/pretty.js";
import { Level } from "./log/level.js";
import { stdout } from "./stream.js";

export type PrintOutput<ToString extends boolean | undefined> = ToString extends undefined
  ? undefined
  : ToString extends false
    ? undefined
    : ToString extends true
      ? string
      : never;

export type PrintOptions<ToString extends boolean> = {
  /**
   * Whether to add an empty line above the content.
   *
   * @default false
   */
  padTop?: boolean;

  /**
   * Whether to add a new line after the content.
   *
   * @default true
   */
  addNewLine?: boolean;

  /**
   * Whether to return the formatted string instead of printing it.
   *
   * @default false
   */
  toStr?: ToString;

  /**
   * What to print if --json was passed.
   *
   * @default undefined (print nothing)
   */
  json?: Record<string, Field>;

  /**
   * The options to pass to `boxen`.
   *
   * @default undefined (no box)
   */
  boxen?: BoxenOptions;
};

export type print<ToString extends boolean, Options extends PrintOptions<ToString>> = {
  /**
   * Prints the given string with dedent.
   *
   * @param str - The string to dedent.
   * @example
   * print("Hello, world!");
   * // => "Hello, world!"
   *
   * print(`
   *   Hello, world!
   *
   *   How are you?
   * `);
   * // => "Hello, world!\n\nHow are you?"
   * @see dedent https://github.com/tamino-martinius/node-ts-dedent
   */
  (str: string): PrintOutput<Options["toStr"]>;

  /**
   * Prints the given template string with dedent and chalk-template.
   *
   * @param template - The template string to format.
   * @param values - The values to interpolate into the template.
   * @example
   * ```
   * let name = "Jane";
   *
   * print`Hello, ${name}!`;
   * // => "Hello, Jane!"
   *
   * print`Hello, {red ${name}}!`;
   * // => "Hello, \u001b[31mJane\u001b[39m!"
   *
   * print`
   *   Hello, {red ${name}}!
   *
   *   How are you?
   * `;
   * // => "Hello, \u001b[31mJane\u001b[39m!\n\nHow are you?"
   * ```
   * @see dedent https://github.com/tamino-martinius/node-ts-dedent
   * @see chalk-template https://github.com/chalk/chalk-template
   */
  (template: TemplateStringsArray, ...values: unknown[]): PrintOutput<Options["toStr"]>;

  /**
   * Configures print with options before printing the given template
   * string with dedent and chalk-template.
   *
   * @example
   * ```
   * let name = "Jane";
   * print({ marginTop: true })`Hello, ${name}!`;
   * // => "\nHello, Jane!"
   *
   * print({ marginTop: true })`Hello, {red ${name}}!`;
   * // => "\nHello, \u001b[31mJane\u001b[39m!"
   *
   * print({ marginTop: true })`
   *   Hello, {red ${name}}!
   *
   *   How are you?
   * `;
   * // => "\nHello, \u001b[31mJane\u001b[39m!\n\nHow are you?"
   * ```
   * @see PrintOptions
   */
  // eslint-disable-next-line @typescript-eslint/prefer-function-type
  <const ToString extends boolean, const NewOptions extends PrintOptions<ToString>>(
    options: PrintOptions<ToString>,
  ): print<ToString, Options & NewOptions>;
};

export const createPrint = <const ToString extends boolean, const Options extends PrintOptions<ToString>>(
  options: Options,
): print<ToString, Options> => {
  const print = ((
    templateOrOptions: PrintOptions<ToString> | string | TemplateStringsArray,
    ...values: unknown[]
  ): print<ToString, PrintOptions<ToString>> | string | undefined => {
    if (!isString(templateOrOptions) && !Array.isArray(templateOrOptions)) {
      return createPrint({ ...options, ...templateOrOptions });
    }

    options = defaults(options, {});

    if (config.logFormat === "json") {
      if (options.json) {
        stdout.write(JSON.stringify(options.json) + "\n");
      }
      return undefined;
    }

    let content = templateOrOptions as string | TemplateStringsArray;
    if (!isString(content)) {
      content = dedent(chalkTemplate(content, ...values));
    }

    if (options.boxen) {
      content = boxen(content, options.boxen);
    }

    if ((options.padTop ?? false) && !content.startsWith("\n")) {
      content = "\n" + content;
    }

    if ((options.addNewLine ?? true) && !content.endsWith("\n")) {
      content += "\n";
    }

    if (options.toStr ?? false) {
      return content;
    }

    if (config.logLevel < Level.PRINT) {
      content = formatPretty(Level.PRINT, "", content, {});
    }

    stdout.write(content);

    return undefined;
  }) as print<ToString, Options>;

  return print;
};

export const print = createPrint({ addNewLine: false });
export const println = createPrint({ addNewLine: true });

export const sprint = createPrint({ toStr: true, addNewLine: false });
export const sprintln = createPrint({ toStr: true, addNewLine: true });

export const printTable = <const ToString extends boolean>({
  message,
  headers,
  rows,
  footer,
  borders: borderType = "none",
  spaceY = 0,
  colAligns = [],
  colWidths = [],
  ...printOptions
}: PrintTableOptions<ToString>): PrintOutput<ToString> => {
  const table = new CliTable3({
    chars: borders[borderType],
    colAligns,
    colWidths,
    head: headers,
    style: { head: [], border: [] },
  });

  table.push(...rows);

  const padding = "\n".repeat(spaceY + 1);

  let output = "";
  if (message) {
    output += message + padding;
  }

  if (borderType === "none") {
    // remove the left padding
    output += dedent(table.toString()).slice(1);
  } else {
    output += table.toString();
  }

  if (footer) {
    output += padding + footer;
  }

  return createPrint(printOptions)(output) as PrintOutput<ToString>;
};

export type PrintTableOptions<ToString extends boolean> = PrintOptions<ToString> & {
  /**
   * The message to print above the table.
   */
  message?: string;

  /**
   * The headers of the table.
   */
  headers?: string[];

  /**
   * The rows of the table.
   */
  rows: string[][];

  /**
   * The message to print below the table.
   */
  footer?: string;

  /**
   * The type of borders to use.
   *
   * @default "none"
   */
  borders?: "none" | "thin" | "thick";

  /**
   * The amount of empty lines to print between the message, table,
   * and footer.
   *
   * @default 0
   */
  spaceY?: number;

  /**
   * The alignment of the content in each column.
   *
   * @default [] (left-aligned)
   */
  colAligns?: ("left" | "center" | "right")[];

  /**
   * The width of each column.
   *
   * @default [] (auto-sized)
   */
  colWidths?: number[];
};

// prettier-ignore
const borders = {
  none: {
       "top-left": "",    top: "",    "top-mid": "",    "top-right": "",
       "left-mid": "",    mid: "",    "mid-mid": "",    "right-mid": "",
             left: "",                   middle: "",          right: "",
    "bottom-left": "", bottom: "", "bottom-mid": "", "bottom-right": "",
  },
  thin: {
       "top-left": "┌",    top: "─",    "top-mid": "┬",    "top-right": "┐",
       "left-mid": "├",    mid: "─",    "mid-mid": "┼",    "right-mid": "┤",
             left: "│",                    middle: "│",          right: "│",
    "bottom-left": "└", bottom: "─", "bottom-mid": "┴", "bottom-right": "┘",
  },
  thick: {
       "top-left": "╔",    top: "═",    "top-mid": "╤",    "top-right": "╗",
             left: "║",                    middle: "│",          right: "║",
       "left-mid": "╟",    mid: "─",    "mid-mid": "┼",    "right-mid": "╢",
    "bottom-left": "╚", bottom: "═", "bottom-mid": "╧", "bottom-right": "╝",
  },
};
