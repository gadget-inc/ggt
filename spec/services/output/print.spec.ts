import chalk from "chalk";
import { describe, expect, it, vi } from "vitest";
import { print, printTable, sprint, sprintln } from "../../../src/services/output/print.js";
import { withEnv } from "../../__support__/env.js";
import { expectStdout } from "../../__support__/stream.js";

describe("print", () => {
  it("writes to stdout", () => {
    print("Hello, world!");
    expectStdout().toMatchInlineSnapshot('"Hello, world!"');
  });

  it("strips indentation from multiline strings", () => {
    print(`
        Hello, world!
        I am a multiline string.
    `);
    expectStdout().toMatchInlineSnapshot(`
      "Hello, world!
      I am a multiline string."
    `);
  });

  it("supports tagged template literals", () => {
    print`
        Hello, world!
        I am a multiline string.
    `;
    expectStdout().toMatchInlineSnapshot(`
      "Hello, world!
      I am a multiline string."
    `);
  });

  it("supports top padding", () => {
    print({ padTop: true })("Hello, world!");
    expectStdout().toMatchInlineSnapshot(`
      "
      Hello, world!"
    `);
  });

  it("supports chalk colors within template literals", () => {
    try {
      chalk.level = 3;
      print`
        {bold Hello, world!}
        {underline I am a multiline string.}
      `;
      expectStdout().toMatchInlineSnapshot(`
        "[1mHello, world![22m
        [4mI am a multiline string.[24m"
      `);
    } finally {
      chalk.level = 0;
    }
  });

  it("prints like a structured logger when GGT_LOG_LEVEL is set", () => {
    try {
      vi.useFakeTimers();
      vi.setSystemTime(0);

      withEnv({ GGT_LOG_LEVEL: "info" }, () => {
        print("Hello, world!");
        expectStdout().toMatchInlineSnapshot(`
          "12:00:00 PRINT : Hello, world!
          "
        `);
      });
    } finally {
      vi.useRealTimers();
    }
  });

  it("prints json when GGT_LOG_FORMAT=json", () => {
    withEnv({ GGT_LOG_FORMAT: "json" }, () => {
      print({ json: { hello: "world" } })("Hello, world!");
      expectStdout().toMatchInlineSnapshot('"{\\"hello\\":\\"world\\"}"');
    });
  });
});

describe("printTable", () => {
  it("writes a table to stdout", () => {
    printTable({
      message: "Table Title",
      headers: ["Column 1", "Column 2"],
      rows: [
        ["Row 1, Column 1", "Row 1, Column 2"],
        ["Row 2, Column 1", "Row 2, Column 2"],
      ],
      footer: "Table Footer",
    });
    expectStdout().toMatchSnapshot();
  });

  it("writes a table with without a footer to stdout", () => {
    printTable({
      message: "Table Title",
      headers: ["Column 1", "Column 2"],
      rows: [
        ["Row 1, Column 1", "Row 1, Column 2"],
        ["Row 2, Column 1", "Row 2, Column 2"],
      ],
    });
    expectStdout().toMatchSnapshot();
  });

  //   it("writes the table as json when GGT_LOG_FORMAT=json", () => {
  //     withEnv({ GGT_LOG_FORMAT: "json" }, () => {
  //       printTable({
  //         message: "Table Title",
  //         headers: ["Column 1", "Column 2"],
  //         rows: [
  //           ["Row 1, Column 1", "Row 1, Column 2"],
  //           ["Row 2, Column 1", "Row 2, Column 2"],
  //         ],
  //         footer: "Table Footer",
  //       });
  //       expectStdout().toMatchSnapshot();
  //     });
  //   });

  for (const borders of ["none", "thin", "thick"] as const) {
    it(`writes a table with ${borders} borders to stdout`, () => {
      printTable({
        borders,
        message: "Table Title",
        headers: ["Column 1", "Column 2"],
        rows: [
          ["Row 1, Column 1", "Row 1, Column 2"],
          ["Row 2, Column 1", "Row 2, Column 2"],
        ],
        footer: "Table Footer",
      });
      expectStdout().toMatchSnapshot();
    });
  }
});

describe("sprint", () => {
  it("accepts a string", () => {
    const result = sprint("hello");
    expect(result).toBe("hello");
  });

  it("accepts a template", () => {
    const world = "world";
    const result = sprint`hello ${world}`;
    expect(result).toBe("hello world");
  });

  it("adds top padding", () => {
    const result = sprint({ padTop: true })("hello");
    expect(result).toBe("\nhello");
  });
});

describe("sprintln", () => {
  it("accepts a string", () => {
    const result = sprintln("hello");
    expect(result).toBe("hello\n");
  });

  it("accepts a template", () => {
    const world = "world";
    const result = sprintln`hello ${world}`;
    expect(result).toBe("hello world\n");
  });

  it("adds top padding", () => {
    const result = sprintln({ padTop: true })("hello");
    expect(result).toBe("\nhello\n");
  });
});
