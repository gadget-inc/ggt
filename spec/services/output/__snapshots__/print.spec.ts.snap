// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`printTable > writes a table to stdout 1`] = `
"Table Title
Column 1         Column 2        
Row 1, Column 1  Row 1, Column 2 
Row 2, Column 1  Row 2, Column 2 
Table Footer
"
`;

exports[`printTable > writes a table with none borders to stdout 1`] = `
"Table Title
Column 1         Column 2        
Row 1, Column 1  Row 1, Column 2 
Row 2, Column 1  Row 2, Column 2 
Table Footer
"
`;

exports[`printTable > writes a table with thick borders to stdout 1`] = `
"Table Title
╔═════════════════╤═════════════════╗
║ Column 1        │ Column 2        ║
╟─────────────────┼─────────────────╢
║ Row 1, Column 1 │ Row 1, Column 2 ║
╟─────────────────┼─────────────────╢
║ Row 2, Column 1 │ Row 2, Column 2 ║
╚═════════════════╧═════════════════╝
Table Footer
"
`;

exports[`printTable > writes a table with thin borders to stdout 1`] = `
"Table Title
┌─────────────────┬─────────────────┐
│ Column 1        │ Column 2        │
├─────────────────┼─────────────────┤
│ Row 1, Column 1 │ Row 1, Column 2 │
├─────────────────┼─────────────────┤
│ Row 2, Column 1 │ Row 2, Column 2 │
└─────────────────┴─────────────────┘
Table Footer
"
`;

exports[`printTable > writes a table with without a footer to stdout 1`] = `
"Table Title
Column 1         Column 2        
Row 1, Column 1  Row 1, Column 2 
Row 2, Column 1  Row 2, Column 2 
"
`;
